src/ppx/scene/scene_material.cpp

```diff
@@ -0,0 +1,217 @@
+{
+    scene::VertexAttributeFlags attrFlags = scene::VertexAttributeFlags::None();
+    attrFlags.bits.texCoords              = true;
+    return attrFlags;
+}
+
+void UnlitMaterial::SetBaseColorFactor(const float4& value)
+{
+    mBaseColorFactor = value;
+}
+
+// -------------------------------------------------------------------------------------------------
+// StandardMaterial
+// -------------------------------------------------------------------------------------------------
+scene::VertexAttributeFlags StandardMaterial::GetRequiredVertexAttributes() const
+{
+    scene::VertexAttributeFlags attrFlags = scene::VertexAttributeFlags::None();
+    attrFlags.bits.texCoords              = true;
+    attrFlags.bits.normals                = true;
+    attrFlags.bits.tangents               = true;
+    attrFlags.bits.colors                 = true;
+    return attrFlags;
+}
+
+void StandardMaterial::SetBaseColorFactor(const float4& value)
+{
+    mBaseColorFactor = value;
+}
+
+void StandardMaterial::SetMetallicFactor(float value)
+{
+    mMetallicFactor = value;
+}
+
+void StandardMaterial::SetRoughnessFactor(float value)
+{
+    mRoughnessFactor = value;
+}
+
+void StandardMaterial::SetOcclusionStrength(float value)
+{
+    mOcclusionStrength = value;
+}
+
+void StandardMaterial::SetEmissiveFactor(const float3& value)
+{
+    mEmissiveFactor = value;
+}
+
+void StandardMaterial::SetEmissiveStrength(float value)
+{
+    mEmissiveStrength = value;
+}
+
+// -------------------------------------------------------------------------------------------------
+// MaterialFactory
+// -------------------------------------------------------------------------------------------------
+MaterialFactory::MaterialFactory()
+{
+}
+
+MaterialFactory::~MaterialFactory()
+{
+}
+
+scene::VertexAttributeFlags MaterialFactory::GetRequiredVertexAttributes(const std::string& materialIdent) const
+{
+    scene::VertexAttributeFlags attrFlags = scene::VertexAttributeFlags::None();
+
+    // Error material only uses position and no other vertex attribues.
+    // No need to put in a specific if statement for it.
+    if (materialIdent == PPX_MATERIAL_IDENT_UNLIT) {
+        attrFlags = UnlitMaterial().GetRequiredVertexAttributes();
+    }
+    else if (materialIdent == PPX_MATERIAL_IDENT_STANDARD) {
+        attrFlags = StandardMaterial().GetRequiredVertexAttributes();
+    }
+
+    return attrFlags;
+}
+
+scene::MaterialRef MaterialFactory::CreateMaterial(
+    const std::string& materialIdent) const
+{
+    scene::MaterialRef material;
+
+    if (materialIdent == PPX_MATERIAL_IDENT_UNLIT) {
+        material = scene::MakeRef(new scene::UnlitMaterial());
+    }
+    else if (materialIdent == PPX_MATERIAL_IDENT_STANDARD) {
+        material = scene::MakeRef(new scene::StandardMaterial());
+    }
+    else {
+        if (!mErrorMaterial) {
+            mErrorMaterial = scene::MakeRef(new scene::ErrorMaterial());
```

Comment:
This bit I'm not sure about. We'll need to see how it's used, but for other materials, we return a ref this factory doesn't own anymore. But here, it's still owned.
Because we use shared_ptr, the refcount will save us if the factory gets deleted, but it might feel wrong.

Maybe this caching should be done in the resource manager & the scene loader handle this logic:
 -  iterates over GLTF nodes
 - if it has material: query the resource for the material. If not yet there, gets created.
 - if it has no material: query the RM for the error material, if not yet there, gets created.

Reply:
Changed it so that ErrorMaterial gets a new'd copy like the other materials. I think if we see the ErroMaterial, we'll just fix the material. Don't think it will break anyone's memory budget.